### DISCUSSION TITLE
How to Update a WinForm GUI From a Route?

### DISCUSSION THREAD

**Discussion by dev-serena** [scottoffen/grapevine#65] [Q&A] [answered]
Hello, I know this is not a Grapevine problem, but how can I update a Winform GUI?

```csharp
[RestResource]
public class MyResource
{
    [RestRoute("Get", "/api/test")]
    public async Task Test(IHttpContext context)
    {
        await context.Response.SendResponseAsync("Successfully hit the test route!");
        //UpdateGUI
    }

    private void UpdateGui(string text)
    {
        labelx.Text = text;
    }
}
```

**Comment by scottoffen** [comment, marked as answer]
_You are correct, this is absolutely outside the scope of Grapevine. That being said, I'm sure many people will have this same question, so it's a good idea to address this here. Keep in mind that this is not necessarily the most optimal way to approach doing this, but rather a concise, working example to get you up and running on the right path._

In order to update your WinForms GUI, you have to get a handle to the open form and the specific control you want to update on it.  If, for example, you have a form named `Form1` that contains a text box named `textBox1`, and you want to update the textbox, you would import `System.Windows.Forms` into your route class, and use something like this:

```csharp
var form = Application.OpenForms["Form1"] as Form1;
var tb = form?.Controls["textBox1"] ?? null;
```

If you try to interact with that control directly, you will get an `System.InvalidOperationException` with the message `Cross-thread operation not valid: Control 'textBox1' accessed from a thread other than the thread it was created on.`

Instead, you want to invoke  a delegate:

```csharp
tb?.Invoke(new MethodInvoker(delegate { tb.Text = "I have updated the value in this text box"; }));
```

That should get you headed in the right direction!

**Reply by scottoffen** [reply]
Furthermore, here is the recommended way to add a rest server to your WinForms application, assuming you want the server running throughout your entire application lifecycle.

```csharp
using System;
using System.Windows.Forms;
using Grapevine;

namespace YourApplication
{
    static class Program
    {
        /// <summary>
        ///  The main entry point for the application.
        /// </summary>
        [STAThread]
        static void Main()
        {
            // Configure your server in the Startup class
            var server = RestServerBuilder.From<Startup>().Build();
            server.Start();

            Application.SetHighDpiMode(HighDpiMode.SystemAware);
            Application.EnableVisualStyles();
            Application.SetCompatibleTextRenderingDefault(false);
            Application.Run(new Form1());

            server.Stop();
        }
    }
}
```

Most of that is boilerplate when you create a new WinForm application. The idea is that you create and start your server before the application starts. The line `Application.Run(...)` will block the thread until the form is exited, at which point you can stop your server.

> DO NOT USE `server.Run()` with a WinForms application, as this blocks the thread until the server is stopped.

**Comment by dev-serena** [comment]
Thanks for tips. I have trying another method you give who work good (with v4):

```csharp
private void Form1_Load(object sender, EventArgs e)
{
    mServer = new RestServer();
    mServer.Properties.Add("CurrentForm", this);
}

private void UpdateGuil(string newText)
{
    this.Invoke(new MethodInvoker(() => textBox1.Text = newText));
}
```

and

```csharp
var form = context.Server.GetPropertyValueAs<Form1>("CurrentForm");
form.UpdateGui("message");
````

**Reply by dev-serena** [reply]
Now i need to find how to invoke UpdateGui with Grapevine 5 :)

**Reply by scottoffen** [reply]
I do not understand your response.
